README.md

```diff
@@ -260,7 +260,9 @@ A Service Binding describes the connection between a [Provisioned Service](#prov
 
 Restricting service binding to resources within the same namespace is strongly **RECOMMENDED**.  Implementations that choose to support cross-namespace service binding **SHOULD** provide a security model that prevents attacks like privilege escalation and secret enumeration, as well as a deterministic way to declare target namespaces.
 
-A Service Binding resource **MUST** define a `.spec.application` which is an `ObjectReference`-like declaration.  A `ServiceBinding` **MAY** define the application reference by-name or by-[label selector][ls]. A name and selector **MUST NOT** be defined in the same reference.  A Service Binding resource **MUST** define a `.spec.service` which is an `ObjectReference`-like declaration to a Provisioned Service-able resource.  Extensions and implementations **MAY** allow additional kinds of applications and services to be referenced.
+A Service Binding resource **MUST** define a `.spec.application` which is an `ObjectReference`-like declaration.  A `ServiceBinding` **MAY** define the application reference by-name or by-[label selector][ls]. A name and selector **MUST NOT** be defined in the same reference.  A Service Binding resource **MUST** define a `.spec.service` which is an `ObjectReference`-like declaration to a Provisioned Service-able resource.  A `ServiceBinding` **MAY** define the service reference by-name or by-[label selector][ls].  Extensions and implementations **MAY** allow additional kinds of applications and services to be referenced.  
+
+As label selectors are inherently queries that return zero-to-many resources, the reconciler implementation **MUST** select the first match and ignore any other matches.  Therefore it is **RECOMMENDED** that `ServiceBinding` authors use a combination of labels that yield a single resource.  
```

Comment:
The main point here is that the spec is allowing for selectors to be used in the cases where the `name` is not known a head of time, but it is strongly recommending that an appropriate set of labels be used so that 1 and only 1 resource matches.  I think that's a reasonable comprise to allow for the dynamic name use case while still keeping a 1-to-1 cardinality between app - Service Binding - service.

Reply:
> but it is strongly recommending that an appropriate set of labels be used so that 1 and only 1 resource matches

The proposed change is not a strong recommendation, but a new requirement that changes the existing behavior.

Current text:

> implementors **MUST** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource.

Proposed text:

> the reconciler implementation **MUST** select the first match and ignore any other matches.

Reply:
From an implementors perspective, I'm not sure what "first match" means. Can you elaborate on the desired semantics?

Reply:
> The proposed change is not a strong recommendation, but a new requirement that changes the existing behavior.

There are two personas: the author of the ServiceBinding CR (let's call "the user") and the implementation of the spec.  There's a strong recommendation for the user to use a combination of labels to yield a single match, because the implementation has a requirement to pick the first one it finds.

I think we want to leave the matching order as undefined, hence the strong recommendation to the user.

Reply:
The proposed text is specifically calling out that "reconciler implementation **MUST**" do something, but that something is ambiguous. The word "first" is troubling me because k8s label selectors don't have an order. We're asking implementations to distinguish between two things without giving them any guidance on how to make that distinction.

Reply:
I've now updated the PR to be a bit more descriptive, let me know if this helps:


> As label selectors are inherently queries that return zero-to-many resources, the reconciler implementation **MUST** select whichever resource it first matched and ignore any other matches.  Given that ordering of resource matching is undefined, it is **RECOMMENDED** that `ServiceBinding` authors use a combination of labels that yield a single resource.